oracles/parallelism/src/main/java/de/learnlib/oracle/parallelism/StaticParallelOracleBuilder.java

```diff
@@ -117,15 +115,17 @@ public StaticParallelOracleBuilder(Supplier<? extends MembershipOracle<I, D>> or
 
     @Nonnull
     public StaticParallelOracle<I, D> createAny() {
-        return new StaticParallelOracle(create(), minBatchSize, poolPolicy);
+        return new StaticParallelOracle<>(create(), minBatchSize, poolPolicy);
     }
 
     @Nonnull
+    @SuppressWarnings("unchecked")
     public DFAStaticParallelOracle<I> createDFA() {
         return new DFAStaticParallelOracle(create(), minBatchSize, poolPolicy);
     }
 
     @Nonnull
+    @SuppressWarnings("unchecked")
     public <O> MealyStaticParallelOracle<I, O> createMealy() {
         return new MealyStaticParallelOracle(create(), minBatchSize, poolPolicy);
```

Comment:
I tried, but then it says it can not infer the type arguments. The reason I needed these specializations is because of these constructors: https://github.com/Meijuh/learnlib/blob/master/oracles/emptiness-oracles/src/main/java/de/learnlib/oracle/emptiness/BreadthFirstEmptinessOracle.java#L53-L62, however, looking at the constructors here: https://github.com/Meijuh/learnlib/blob/master/oracles/equivalence-oracles/src/main/java/de/learnlib/oracle/equivalence/WMethodEQOracle.java#L86-L102 the `DFAMembershipOracle` and `MealyMembershipOracle` are not used. Is there any use for `DFAMembershipOracle` and `MealyMembershipOracle`?

Do you know how to property construct a MealyStaticParallelOracle, and DFAStaticParallelOracle?

Reply:
The DFA- and Mealy- specializations (e.g. for EQOracles) are mainly just convenience for programming, so that you don't always have to write `EquivalenceOracle<MealyMachine<?, I, ?, O>, ...`. However, the inheritance only works one way. If you require e.g. a `MembershipOracle<I, Boolean>` you can pass in a `DFAMembershipOracle<I>`, because it extends the generic MO. But if you require a `DFAMembershipOracle<I>` you cannot pass a `MembershipOracle<I, Boolean>`. Thus it is best for parameters to always choose the broadest type possible.

I think, if you change the constructors of the `DFABFEmptinessOracle`, etc. to require a `MembershipOracle<I, Boolean>` rather than a `DFAMembershipOracle<I>` (similar to e.g. `DFAWMethodEQOracle`) your problem will resolve: People still can pass a `DFAMembershipOracle<I>` to the constructor, but every other fitting implementation (essentially a `MembershipOracle<I, Boolean>`) that happens to not implement `DFAMembershipOracle<I>` will work as well.

I think, what you try is a little bit too forced. You would have to write a delegation oracle, that implements `DFAMembershipOracle<I>` and takes a `MembershipOracle<I, Boolean>` as a delegate, to basically simulate the cyclic inheritance you want. With the suggested changes above, I think the previous version of the `StaticParallelOracleBuilder`(where `create()` simply returned a `StaticParallelOracle<I, D>`) was just fine, so you don't have to go through all that hassle.

Reply:
Okay I shall undo the additions for the `StaticParallelOracle` and change the constructors fore the EmptinessOracle and InclusionOracle.